File: java/base/src/main/vsdk/toolkit/environment/light/AmbientLight.java
package vsdk.toolkit.environment.light;

import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.environment.geometry.element.Ray;

public final class AmbientLight extends Light
{
    public AmbientLight(ColorRgb emission)
    {
        super(new Vector3Dd(0, 0, 0), emission);
    }

    @Override
    public boolean isAmbient()
    {
        return true;
    }

    @Override
    public LightDirection getDirectionAndDistance(Vector3Dd surfacePoint)
    {
        // Never sampled: shaders resolve the ambient contribution from
        // getEmission() directly and skip the shadow ray entirely.
        return new LightDirection(new Vector3Dd(0, 0, 0), 0.0);
    }

    @Override
    public double evaluateLightResponseFactor(Ray lightSourceRay)
    {
        return 0.0;
    }

    @Override
    public Light copy()
    {
        AmbientLight copy = new AmbientLight(getEmission());
        copy.setId(getId());
        copy.setName(getName());
        return copy;
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/light/AmbientLight.ts
import { ColorRgb } from "../../common/color/ColorRgb.js";
import { Vector3Dd } from "../../common/linealAlgebra/Vector3Dd.js";
import type { Ray } from "../geometry/element/Ray.js";
import { Light } from "./Light.js";

/** Position-less light whose emission is added to every shaded point. */
export class AmbientLight extends Light {
    public constructor(emission: ColorRgb) {
        super(new Vector3Dd(0, 0, 0), emission);
    }

    public override isAmbient(): boolean {
        return true;
    }

    public override getDirectionAndDistance(_surfacePoint: Vector3Dd): Light.LightDirection {
        // Never sampled: shaders resolve the ambient contribution from
        // getEmission() directly and skip the shadow ray entirely.
        return new Light.LightDirection(new Vector3Dd(0, 0, 0), 0.0);
    }

    public override evaluateLightResponseFactor(_lightSourceRay: Ray): number {
        return 0.0;
    }

    public override copy(): Light {
        const copy = new AmbientLight(this.getEmission());
        copy.setId(this.getId());
        copy.setName(this.getName());
        return copy;
    }
}
